spring-integration-mail/src/main/java/org/springframework/integration/mail/AbstractMailReceiver.java

```diff
@@ -298,28 +337,84 @@ private Folder obtainFolderInstance() throws MessagingException {
 					this.logger.debug("Received " + messages.length + " messages");
 				}
 
-				Message[] filteredMessages = filterMessagesThruSelector(messages);
+				MimeMessage[] filteredMessages = filterMessagesThruSelector(messages);
 
 				postProcessFilteredMessages(filteredMessages);
 
-				return filteredMessages;
+				if (this.headerMapper != null) {
+					org.springframework.messaging.Message<?>[] converted =
+							new org.springframework.messaging.Message<?>[filteredMessages.length];
+					int n = 0;
+					for (MimeMessage message : filteredMessages) {
+						Map<String, Object> headers = this.headerMapper.toHeaders(message);
+						converted[n++] = getMessageBuilderFactory().withPayload(extractContent(message, headers))
+								.copyHeaders(headers)
+								.build();
+					}
+					return converted;
+				}
+				else {
+					return filteredMessages;
+				}
 			}
 			finally {
 				MailTransportUtils.closeFolder(this.folder, this.shouldDeleteMessages);
 			}
 		}
 	}
 
+	private Object extractContent(MimeMessage message, Map<String, Object> headers) {
+		Object content;
+		try {
+			content = message.getContent();
+			if (content instanceof String) {
+				String mailContentType = (String) headers.get(MailHeaders.CONTENT_TYPE);
```

Comment:
Something has been missed here...
I don't see the `MailHeaders.CONTENT_TYPE` in the `DefaultMailHeaderMapper`.
Just a simple omission? 

Let me know and I'll fix on merge!